search/query.go

```diff
@@ -114,7 +121,45 @@ func ParseQuery(q string) Query {
 	return Query{Term: strings.Join(terms, " "), ExactTerms: exactTerms, Original: q, Filters: filters}
 }
 
-func createResultsQuery(resultTypes []string, q Query, docIds []string, filterOutCUSources []string) elastic.Query {
+func createSpanNearQuery(field string, term string, boost float32, slop int, inOrder bool) elastic.Query {
+	clauses := make([]string, 0)
+	spanNearMask := "{\"span_near\": { \"clauses\": [%s], \"slop\": %d, \"boost\": %f, \"in_order\": %t } }"
+	clauseMask := "{\"span_multi\": { \"match\": { \"fuzzy\": { \"%s\": { \"value\": \"%s\", \"fuzziness\": %s, \"transpositions\": %s } } } } }"
+	splitted := strings.Fields(term)
+	for _, t := range splitted {
+		if t == "<" || t == ">" || t == "-" {
+			continue
+		}
+		fuzzines := "\"AUTO\""   // Default.
+		transpositions := "true" // Default.
+		runes := []rune(t)
+		_, convertToIntErr := strconv.Atoi(t)
+		if convertToIntErr == nil || (len(runes) == 3 && runes[1] == '"') || (len(runes) == 4 && runes[2] == '"') {
+			//  We dont use fuzzines for numeric values (number or hebrew numeric representation like מ"ג)
+			fuzzines = "0"
+		} else if len(runes) == 1 && runes[0] >= 'א' && runes[0] <= 'ת' {
+			// This logic allows finding single hebrew letter with ' symbol without the mention of the ' symbol.
+			// The solution is not perfect for all times. In some (rare) cases the letter may be replaced with another latter.
+			fuzzines = "1"
+			transpositions = "false" // Limit the fuzzines not to include transpositions of two adjacent characters (ח' -> 'ח). Maybe not required.
+		}
+		b, err := json.Marshal(t)
+		if err != nil {
+			panic(err) //TBD log
+		}
+		// Trim the beginning and trailing " character
+		esc := string(b[1 : len(b)-1])
+		clause := fmt.Sprintf(clauseMask, field, esc, fuzzines, transpositions)
+		clauses = append(clauses, clause)
+	}
+	clausesStr := strings.Join(clauses, ",")
+	queryStr := fmt.Sprintf(spanNearMask, clausesStr, slop, boost, inOrder)
+	//fmt.Printf("SpanNear Query: %s\n", queryStr)
+	query := elastic.NewRawStringQuery(queryStr)
+	return query
+}
+
+func createResultsQuery(resultTypes []string, q Query, docIds []string, filterOutCUSources []string, titlesOnly bool) elastic.Query {
```

Comment:
Add general comment with all cases of scores....
Like overview of out scoring function.... if words have different order then .... if words don't exactly match then...

Reply:
Very complex and sensitive case...

How we make sure we don't forget any type of field/title/description/content in any type of search/score?

Consider refactoring....

Reply:
TBD add comment about ALL features of this function